File: backend/src/main/java/io/visualzpl/batch/BatchZpl.java
package io.visualzpl.batch;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Server-side mirror of the frontend BatchZpl module.
 *
 * Provides three pure-function entry points so the backend can:
 *   - Extract `{{Variable}}` placeholders from any ZPL template.
 *   - Generate a concatenated batch ZPL stream by performing per-row
 *     substitution against a precompiled base ZPL.
 *   - Validate a grid of input rows against the strictest rule implied by
 *     the consuming object type (Text / Barcode / QR Code).
 *
 * The implementation deliberately stays self-contained — no Spring beans,
 * no IO, no logging — so the unit tests in BatchZplTest can exercise every
 * branch in isolation.
 */
public final class BatchZpl {

    private BatchZpl() {
        // utility class
    }

    /** Matches `{{VariableName}}` tokens. Names must start with a letter or `_`. */
    public static final Pattern VARIABLE_PATTERN =
            Pattern.compile("\\{\\{([A-Za-z_][A-Za-z0-9_]*)}}");

    /** Discriminates the label-object type that consumes a variable. */
    public enum ConsumerType {
        TEXT, BARCODE, QRCODE
    }

    /** Severity of a single validation issue. */
    public enum Severity {
        WARNING, ERROR
    }

    /** Stable machine-readable codes shared with the frontend implementation. */
    public enum IssueCode {
        BARCODE_RULE_EXCEPTION,
        EMPTY_REQUIRED_FIELD,
        CONTROL_CHARACTER_WARNING,
        NON_ASCII_CHARACTER
    }

    public record ValidationIssue(
            int rowIndex,
            String variableName,
            Severity severity,
            IssueCode code,
            String message
    ) {}

    public record ValidationResult(
            List<ValidationIssue> issues,
            Map<String, ValidationIssue> byCell,
            int errorCount,
            int warningCount
    ) {}

    private static final int PRINTABLE_ASCII_MIN = 0x20; // space
    private static final int PRINTABLE_ASCII_MAX = 0x7E; // tilde

    /** Stable map key used in {@link ValidationResult#byCell()}. */
    public static String cellKey(int rowIndex, String variableName) {
        return rowIndex + "#" + variableName;
    }

    /**
     * Returns every distinct `{{name}}` placeholder found in `text`, in the
     * order it was first encountered. Useful when the caller needs to render
     * a column header per variable (matches the frontend behavior).
     */
    public static List<String> extractVariables(String text) {
        Set<String> seen = new LinkedHashSet<>();
        Matcher matcher = VARIABLE_PATTERN.matcher(text);
        while (matcher.find()) {
            seen.add(matcher.group(1));
        }
        return List.copyOf(seen);
    }

    /**
     * Mirrors ZplBuilder.escapeFieldData — `^` and `~` are ZPL control
     * prefixes, so any substituted user value containing them is sanitized
     * to a space before reaching the printer.
     */
    private static String escapeForZpl(String value) {
        return value.replaceAll("[\\^~]", " ");
    }

    /**
     * Performs per-row variable substitution on a precompiled base ZPL
     * string. Missing values for a `{{name}}` placeholder are replaced
     * with an empty string. The output joins `^XA…^XZ` blocks with newlines
     * so the firmware advances and cuts one label per block.
     */
    public static String generateBatchZpl(String baseZpl, List<Map<String, String>> rows) {
        if (rows.isEmpty()) {
            return baseZpl;
        }
        StringBuilder out = new StringBuilder(baseZpl.length() * rows.size());
        for (int i = 0; i < rows.size(); i++) {
            Map<String, String> row = rows.get(i);
            Matcher matcher = VARIABLE_PATTERN.matcher(baseZpl);
            StringBuilder rendered = new StringBuilder(baseZpl.length());
            while (matcher.find()) {
                String name = matcher.group(1);
                String value = row.get(name);
                String replacement = value != null ? escapeForZpl(value) : "";
                matcher.appendReplacement(rendered, Matcher.quoteReplacement(replacement));
            }
            matcher.appendTail(rendered);
            if (i > 0) {
                out.append('\n');
            }
            out.append(rendered);
        }
        return out.toString();
    }

    /**
     * Validates every cell of `rows` against the strictest rule implied by
     * the consuming object type:
     *   BARCODE > QRCODE > TEXT  (strictest first)
     *
     * The `consumers` map must list which ConsumerType set each variable
     * appears in. A variable shared between TEXT and BARCODE is held to
     * the BARCODE rule (NON_ASCII_CHARACTER becomes an ERROR rather than
     * the TEXT-level CONTROL_CHARACTER_WARNING).
     */
    public static ValidationResult validateBatchData(
            Map<String, Set<ConsumerType>> consumers,
            List<Map<String, String>> rows
    ) {
        List<ValidationIssue> issues = new ArrayList<>();
        for (int rowIndex = 0; rowIndex < rows.size(); rowIndex++) {
            Map<String, String> row = rows.get(rowIndex);
            for (Map.Entry<String, Set<ConsumerType>> entry : consumers.entrySet()) {
                String name = entry.getKey();
                String value = row.getOrDefault(name, "");
                ValidationIssue issue = validateCellValue(rowIndex, name, value, entry.getValue());
                if (issue != null) {
                    issues.add(issue);
                }
            }
        }

        Map<String, ValidationIssue> byCell = new HashMap<>();
        for (ValidationIssue issue : issues) {
            String key = cellKey(issue.rowIndex(), issue.variableName());
            ValidationIssue existing = byCell.get(key);
            if (existing == null
                    || severityRank(issue.severity()) > severityRank(existing.severity())) {
                byCell.put(key, issue);
            }
        }

        int errorCount = 0;
        int warningCount = 0;
        for (ValidationIssue issue : byCell.values()) {
            if (issue.severity() == Severity.ERROR) {
                errorCount++;
            } else {
                warningCount++;
            }
        }

        return new ValidationResult(
                Collections.unmodifiableList(issues),
                Collections.unmodifiableMap(byCell),
                errorCount,
                warningCount
        );
    }

    private static int severityRank(Severity severity) {
        return severity == Severity.ERROR ? 2 : 1;
    }

    private static ValidationIssue validateCellValue(
            int rowIndex,
            String variableName,
            String value,
            Set<ConsumerType> consumers
    ) {
        if (consumers.contains(ConsumerType.BARCODE)) {
            if (value.isEmpty()) {
                return new ValidationIssue(
                        rowIndex, variableName, Severity.ERROR,
                        IssueCode.EMPTY_REQUIRED_FIELD,
                        "Barcode value cannot be empty.");
            }
            for (int i = 0; i < value.length(); i++) {
                char ch = value.charAt(i);
                int code = ch;
                if (ch == '^' || ch == '~') {
                    return new ValidationIssue(
                            rowIndex, variableName, Severity.ERROR,
                            IssueCode.BARCODE_RULE_EXCEPTION,
                            "ZPL control characters ^ and ~ are not allowed in barcode data.");
                }
                if (code < PRINTABLE_ASCII_MIN || code > PRINTABLE_ASCII_MAX) {
                    return new ValidationIssue(
                            rowIndex, variableName, Severity.ERROR,
                            IssueCode.NON_ASCII_CHARACTER,
                            String.format(
                                    "Code 128 does not support character \"%c\" (U+%04X).",
                                    ch, code));
                }
            }
            return null;
        }

        if (consumers.contains(ConsumerType.QRCODE)) {
            if (value.isEmpty()) {
                return new ValidationIssue(
                        rowIndex, variableName, Severity.WARNING,
                        IssueCode.EMPTY_REQUIRED_FIELD,
                        "QR code value is empty.");
            }
            if (containsControlChar(value)) {
                return new ValidationIssue(
                        rowIndex, variableName, Severity.WARNING,
                        IssueCode.CONTROL_CHARACTER_WARNING,
                        "Control characters ^ and ~ will be replaced with spaces on print.");
            }
            return null;
        }

        // Text — least restrictive: only warn on ZPL control characters.
        if (containsControlChar(value)) {
            return new ValidationIssue(
                    rowIndex, variableName, Severity.WARNING,
                    IssueCode.CONTROL_CHARACTER_WARNING,
                    "Control characters ^ and ~ will be replaced with spaces on print.");
        }
        return null;
    }

    private static boolean containsControlChar(String value) {
        for (int i = 0; i < value.length(); i++) {
            char ch = value.charAt(i);
            if (ch == '^' || ch == '~') {
                return true;
            }
        }
        return false;
    }
}

File: frontend/src/BatchZpl.ts
/**
 * BatchZpl
 *
 * Variable-substitution pipeline that turns a single LabelDocument into a
 * concatenated ZPL stream of N labels, one per data row.
 *
 * Variable syntax: any token enclosed in double curly braces, where the name
 * is a valid identifier — e.g. `{{Price}}`, `{{Item_Code}}`, `{{BarcodeData}}`.
 *
 * Why string-level substitution instead of per-row deep-cloning + rebuild?
 *   1. ZplBuilder.escapeFieldData() replaces `^` and `~` with spaces, but
 *      leaves `{` and `}` alone, so placeholders survive into the final ZPL.
 *   2. Running ZplBuilder once is cheaper than rebuilding for every row.
 *   3. The same escape rule (`escapeForZpl`) is applied to substituted values
 *      so user-supplied data containing `^` or `~` cannot break ZPL parsing.
 */

import { ZplBuilder } from './ZplBuilder';
import type { LabelDocument } from './types';

/** Matches `{{VariableName}}` tokens. Names must start with a letter or `_`. */
export const VARIABLE_PATTERN = /\{\{([A-Za-z_][A-Za-z0-9_]*)\}\}/g;

/** Plain key→value mapping for a single batch row. */
export type DataRow = Record<string, string>;

/**
 * Scan the document for all distinct `{{name}}` placeholders that appear in
 * any field that gets rendered into ZPL (Text / Barcode / QR Code data).
 *
 * Returns names in discovery order so the UI columns stay stable across
 * re-renders and small doc edits.
 */
export function extractVariables(doc: LabelDocument): string[] {
  const seen = new Set<string>();
  const ordered: string[] = [];

  for (const obj of doc.objects) {
    // Image data field is just the filename — it never reaches the printer.
    if (obj.type === 'image') continue;

    // Use a fresh regex each iteration so .lastIndex from a previous match
    // does not leak across object boundaries.
    const pattern = new RegExp(VARIABLE_PATTERN.source, 'g');
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(obj.data)) !== null) {
      const name = match[1];
      if (!seen.has(name)) {
        seen.add(name);
        ordered.push(name);
      }
    }
  }

  return ordered;
}

/**
 * Mirrors ZplBuilder.escapeFieldData so user-entered values can never
 * accidentally introduce a ZPL control prefix (`^` or `~`).
 */
function escapeForZpl(value: string): string {
  return value.replace(/[\^~]/g, ' ');
}

/**
 * Compile the document once, then perform per-row string substitution on the
 * resulting ZPL. The output is a single ZPL stream containing N `^XA…^XZ`
 * blocks back to back — Zebra firmware advances and cuts one label per block.
 *
 * If a row is missing a value for some variable, the placeholder is replaced
 * with an empty string (i.e. printed as blank in that field).
 *
 * If `dataRows` is empty the base ZPL is returned unchanged so the caller can
 * still pre-flight what a single label looks like.
 */
export function generateBatchZpl(
  doc: LabelDocument,
  dataRows: DataRow[],
): string {
  const baseZpl = new ZplBuilder(doc).build();
  if (dataRows.length === 0) return baseZpl;

  const labels: string[] = new Array(dataRows.length);
  for (let i = 0; i < dataRows.length; i++) {
    const row = dataRows[i];
    const pattern = new RegExp(VARIABLE_PATTERN.source, 'g');
    labels[i] = baseZpl.replace(pattern, (_full, name: string) => {
      const value = row[name];
      return value !== undefined ? escapeForZpl(value) : '';
    });
  }
  return labels.join('\n');
}

// ──────────────────────────────────────────────────────────────────────────
// Batch data validation
// ──────────────────────────────────────────────────────────────────────────

/** Severity of a single validation issue. */
export type ValidationSeverity = 'warning' | 'error';

/** Machine-readable codes shared between UI highlighting and any logging. */
export type ValidationCode =
  | 'BarcodeRuleException'
  | 'EmptyRequiredField'
  | 'ControlCharacterWarning'
  | 'NonAsciiCharacter';

export interface BatchValidationIssue {
  rowIndex: number;
  variableName: string;
  severity: ValidationSeverity;
  code: ValidationCode;
  /** Human-readable message suitable for cell tooltips and accessible alerts. */
  message: string;
}

export interface BatchValidationResult {
  issues: BatchValidationIssue[];
  /** Map keyed by `cellKey(rowIndex, name)` — holds the worst issue per cell. */
  byCell: Map<string, BatchValidationIssue>;
  errorCount: number;
  warningCount: number;
}

/** Stable lookup key used in `byCell`. */
export function cellKey(rowIndex: number, variableName: string): string {
  return `${rowIndex}#${variableName}`;
}

/** Thrown by callers that want a hard failure instead of a soft issue list. */
export class ValidationError extends Error {
  readonly issues: BatchValidationIssue[];
  constructor(issues: BatchValidationIssue[]) {
    super(`Batch validation failed with ${issues.length} issue(s).`);
    this.name = 'ValidationError';
    this.issues = issues;
  }
}

/** Thrown when a value violates Code 128 character constraints. */
export class BarcodeRuleException extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'BarcodeRuleException';
  }
}

const PRINTABLE_ASCII_MIN = 0x20; // space
const PRINTABLE_ASCII_MAX = 0x7e; // tilde

/**
 * Validate every cell of `dataRows` against the constraints implied by the
 * label objects that consume each variable. The strictest object type wins —
 * a variable used in both a Text and a Barcode field is held to the Barcode
 * rules. Returns a structured result the UI can use to highlight cells.
 */
export function validateBatchData(
  doc: LabelDocument,
  dataRows: DataRow[],
): BatchValidationResult {
  type ConsumerType = 'text' | 'barcode' | 'qrcode';

  // Step 1 — classify every variable by the object types that consume it.
  const consumersByVariable = new Map<string, Set<ConsumerType>>();
  for (const obj of doc.objects) {
    if (obj.type === 'image' || obj.type === 'table') continue;
    const pattern = new RegExp(VARIABLE_PATTERN.source, 'g');
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(obj.data)) !== null) {
      const name = match[1];
      if (!consumersByVariable.has(name)) {
        consumersByVariable.set(name, new Set());
      }
      consumersByVariable.get(name)!.add(obj.type);
    }
  }

  // Step 2 — per-cell rule application.
  const issues: BatchValidationIssue[] = [];
  for (let rowIndex = 0; rowIndex < dataRows.length; rowIndex++) {
    const row = dataRows[rowIndex];
    for (const [variableName, consumers] of consumersByVariable) {
      const value = row[variableName] ?? '';
      const issue = validateCellValue(
        rowIndex,
        variableName,
        value,
        consumers,
      );
      if (issue) issues.push(issue);
    }
  }

  // Step 3 — dedupe by cell, keep the worst severity.
  const byCell = new Map<string, BatchValidationIssue>();
  for (const issue of issues) {
    const key = cellKey(issue.rowIndex, issue.variableName);
    const existing = byCell.get(key);
    if (
      !existing ||
      severityRank(issue.severity) > severityRank(existing.severity)
    ) {
      byCell.set(key, issue);
    }
  }

  let errorCount = 0;
  let warningCount = 0;
  for (const issue of byCell.values()) {
    if (issue.severity === 'error') errorCount++;
    else warningCount++;
  }

  return { issues, byCell, errorCount, warningCount };
}

function severityRank(severity: ValidationSeverity): number {
  return severity === 'error' ? 2 : 1;
}

function validateCellValue(
  rowIndex: number,
  variableName: string,
  value: string,
  consumers: Set<'text' | 'barcode' | 'qrcode'>,
): BatchValidationIssue | null {
  // Strictest rule wins — barcode > qrcode > text.
  if (consumers.has('barcode')) {
    if (value.length === 0) {
      return {
        rowIndex,
        variableName,
        severity: 'error',
        code: 'EmptyRequiredField',
        message: 'Barcode value cannot be empty.',
      };
    }
    for (let i = 0; i < value.length; i++) {
      const ch = value[i];
      const code = value.charCodeAt(i);
      if (ch === '^' || ch === '~') {
        return {
          rowIndex,
          variableName,
          severity: 'error',
          code: 'BarcodeRuleException',
          message:
            'ZPL control characters ^ and ~ are not allowed in barcode data.',
        };
      }
      if (code < PRINTABLE_ASCII_MIN || code > PRINTABLE_ASCII_MAX) {
        const hex = code.toString(16).padStart(4, '0').toUpperCase();
        return {
          rowIndex,
          variableName,
          severity: 'error',
          code: 'NonAsciiCharacter',
          message: `Code 128 does not support character "${ch}" (U+${hex}).`,
        };
      }
    }
    return null;
  }

  if (consumers.has('qrcode')) {
    if (value.length === 0) {
      return {
        rowIndex,
        variableName,
        severity: 'warning',
        code: 'EmptyRequiredField',
        message: 'QR code value is empty.',
      };
    }
    if (/[\^~]/.test(value)) {
      return {
        rowIndex,
        variableName,
        severity: 'warning',
        code: 'ControlCharacterWarning',
        message:
          'Control characters ^ and ~ will be replaced with spaces on print.',
      };
    }
    return null;
  }

  // Text — most permissive.
  if (/[\^~]/.test(value)) {
    return {
      rowIndex,
      variableName,
      severity: 'warning',
      code: 'ControlCharacterWarning',
      message:
        'Control characters ^ and ~ will be replaced with spaces on print.',
    };
  }
  return null;
}
